ggml/src/ggml-vulkan.cpp

```diff
@@ -5921,23 +5951,16 @@ static bool ggml_vk_compute_forward(ggml_backend_vk_context * ctx, ggml_tensor *
             memcpy(cpy.dst, cpy.src, cpy.n);
         }
 
-#ifdef GGML_VULKAN_PERF
-        start = std::chrono::steady_clock::now();
-#endif // GGML_VULKAN_PERF
-
-        ggml_vk_submit(subctx, ctx->fence);
+        ggml_vk_submit(subctx, fence);
     }
 
-    if (tensor_idx == subctx->exit_tensor_idx) {
+    if (use_fence) {
         VK_CHECK(ctx->device->device.waitForFences({ ctx->fence }, true, UINT64_MAX), "ggml_vk_compute_forward waitForFences");
 
-#ifdef GGML_VULKAN_PERF
-        auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - start);
-        ctx->device->perf_logger->log_timing(tensor, duration.count());
-#endif // GGML_VULKAN_PERF
-
         ctx->device->device.resetFences({ ctx->fence });
+    }
 
+    if (tensor_idx == subctx->exit_tensor_idx) {
```

Comment:
The outgoing staging buffer memcpys should only happen after the GPU operations are completed, which is after the `waitForFences` call. Since you submit work, but don't wait for it to be completed, the memcpys shouldn't be done.

However, this memcpy system might be leftover from when copying data to and back from the GPU was still the task of the backend, not the program.

Might be enough to replace the in_memcpys and out_memcpys execution with an assertion that they are empty.

Reply:
around line 5941 there is
```
    // always wait for the GPU work to be done for the last submit
    if (tensor_idx == subctx->exit_tensor_idx) {
        use_fence = true;
    }
```

if `use_fence` is true the code waits for the fence and after waiting for the fence the memcpy logic is executed. Thus the behaviour is the same as before.